reliability-engineering/terraform/modules/concourse-monitoring/files/dashboards/concourse-slo.json

```diff
@@ -0,0 +1,417 @@
+        "type": "dashboard"
+      }
+    ]
+  },
+  "description": "service level objectives",
+  "editable": true,
+  "gnetId": null,
+  "graphTooltip": 0,
+  "links": [],
+  "panels": [
+    {
+      "datasource": "CloudWatch",
+      "description": "",
+      "fieldConfig": {
+        "defaults": {
+          "custom": {},
+          "mappings": [
+            {
+              "from": "",
+              "id": 0,
+              "operator": "",
+              "text": "",
+              "to": "",
+              "type": 1,
+              "value": ""
+            }
+          ],
+          "thresholds": {
+            "mode": "percentage",
+            "steps": [
+              {
+                "color": "dark-red",
+                "value": null
+              },
+              {
+                "color": "dark-yellow",
+                "value": 99
+              },
+              {
+                "color": "dark-green",
+                "value": 99.9
+              }
+            ]
+          },
+          "unit": "percentunit"
+        },
+        "overrides": []
+      },
+      "gridPos": {
+        "h": 8,
+        "w": 6,
+        "x": 0,
+        "y": 0
+      },
+      "id": 6,
+      "interval": "",
+      "options": {
+        "colorMode": "value",
+        "graphMode": "area",
+        "justifyMode": "auto",
+        "orientation": "auto",
+        "reduceOptions": {
+          "calcs": [
+            "mean"
+          ],
+          "fields": "/^errorRatio$/",
+          "values": false
+        }
+      },
+      "pluginVersion": "7.0.5",
+      "targets": [
+        {
+          "alias": "",
+          "apiMode": "Metrics",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status5xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_ELB_5XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "A",
+          "region": "eu-west-2",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status4xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_ELB_4XX_Count",
+          "namespace": "AWS/ApplicationELB",
+          "period": "300",
+          "refId": "B",
+          "region": "default",
+          "statistics": [
+            "Sum"
+          ]
+        },
+        {
+          "alias": "",
+          "dimensions": {
+            "LoadBalancer": "app/cd-concourse-web/ba5029fcd9b9cde2"
+          },
+          "expression": "",
+          "hide": true,
+          "id": "status2xx",
+          "matchExact": true,
+          "metricName": "HTTPCode_Target_2XX_Count",
```

Comment:
I think `RequestCount` might be more interesting to count - although confusingly it's not all requests, just all requests that a **target** responded with (so a 502/504 from the load balancer doesn't increase RequestCount)

Reply:
Due to how the CloudWatch math expressions work (and how sparse metrics / FILL can only work on the metric itself) I was not able to create an expression that could reliably calculate  ratios using `RequestCount`.

Reply:
I don't understand the problem you describe: in particular, I don't understand how it's a problem for `RequestCount` but not for `HTTPCode_Target_2XX_Count`

Reply:
I can't say I fully understand it either but attempting to do math expressions with a mix of `HTTPCode_ELB_XXX_Count` and `RequestCount` fails ... so you need to sum up the HTTPCode_ELB_XXX_Counts to get the total count.

My guess is that the timeseries for RequestCount does not line up at the same data points as for the HTTPCode_ELB_XXX_Count metrics.

Reply:
I tried it and it worked fine for me 🤷  (see slack for URL)

Reply:
continued to have issues where math expressions failed over certain time ranges whenever requestCount was involved with the various status code metrics.

I have simplified the expression to exclude 3xx/4xx anyway, so bringing them back using requestCount wouldn't make much sense. I'm going to chalk this up to CloudwatchWeirdness and move on.